File: src/utils/cfr-fetcher.js
// Lógica pura de fetch + parse del 49 CFR desde la API publica de eCFR.
// No toca DB, no toca archivos — solo descarga, parsea, devuelve secciones.
//
// Lo usan: scripts/fetch-cfr.js (CLI) y src/jobs/cfr-update.js (cron).
// Tests: test/cfr-fetcher.test.js

const https = require('https');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Parts 388 y 394 son [Reserved] en el CFR — no tienen contenido.
const PARTS = [380, 381, 382, 383, 384, 385, 386, 387, 389, 390, 391, 392, 393, 395, 396, 397, 398, 399];

const RAW_DIR = path.join(__dirname, '..', '..', 'data', 'imports', 'cfr-raw');

function fetchUrl(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, {
      headers: { Accept: 'application/xml', 'User-Agent': 'BOTDOT-CFR-fetcher/1.0' },
    }, res => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        return fetchUrl(res.headers.location).then(resolve, reject);
      }
      if (res.statusCode !== 200) {
        return reject(new Error(`HTTP ${res.statusCode}: ${url}`));
      }
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
    });
    req.on('error', reject);
    req.setTimeout(30000, () => { req.destroy(new Error('timeout')); });
  });
}

// Pregunta a eCFR cual es el ultimo issue date para Title 49.
async function getLatestIssueDate() {
  const data = await fetchUrl('https://www.ecfr.gov/api/versioner/v1/titles.json');
  const json = JSON.parse(data);
  const t49 = (json.titles || []).find(t => t.number === 49);
  if (!t49 || !t49.latest_issue_date) {
    throw new Error('No se pudo determinar latest_issue_date para Title 49');
  }
  return t49.latest_issue_date;
}

function decodeEntities(s) {
  return s
    .replace(/&#xA7;/gi, '§')
    .replace(/&#xB6;/gi, '¶')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#x([0-9a-f]+);/gi, (_, h) => String.fromCharCode(parseInt(h, 16)))
    .replace(/&#(\d+);/g, (_, n) => String.fromCharCode(parseInt(n, 10)));
}

function stripInlineTags(s) {
  return s.replace(/<\/?[A-Z][A-Z0-9]*( [^>]*)?>/g, '');
}

function parsePartXml(xml, partNum) {
  const sections = [];
  const re = /<DIV8\s+N="([^"]+)"\s+TYPE="SECTION"[^>]*>([\s\S]*?)<\/DIV8>/g;
  let m;
  while ((m = re.exec(xml)) !== null) {
    const sectionNum = m[1];
    const inner = m[2];

    const headMatch = inner.match(/<HEAD>([\s\S]*?)<\/HEAD>/);
    let title = headMatch ? decodeEntities(stripInlineTags(headMatch[1])).trim() : '';
    title = title.replace(/^§\s*[\d.]+\s*/, '').trim();

    const paragraphs = [];
    const pRe = /<P>([\s\S]*?)<\/P>/g;
    let pm;
    while ((pm = pRe.exec(inner)) !== null) {
      paragraphs.push(decodeEntities(stripInlineTags(pm[1])).trim());
    }
    const text = paragraphs.join('\n\n');

    if (!sectionNum) continue;
    sections.push({ section: sectionNum, title, text, part: partNum, keywords: [] });
  }
  return sections;
}

function autoKeywords(title) {
  const STOP = new Set(['for','the','and','with','from','that','this','part','of','to','in','on','a','an','or','by','as','be','is','are','at','it']);
  return (title || '').toLowerCase()
    .replace(/[^a-z0-9 ]/g, ' ')
    .split(/\s+/)
    .filter(t => t.length >= 4 && !STOP.has(t))
    .slice(0, 8);
}

// Hash del contenido normativo (title + text). Usado para detectar cambios.
// Si cambia un solo caracter del texto, el hash cambia.
function hashContent(title, text) {
  return crypto.createHash('sha256').update(title + '|' + text, 'utf8').digest('hex');
}

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

/**
 * Baja todas las Parts 380-399 desde eCFR y devuelve [{section, part, title,
 * text, keywords, content_hash, issue_date}, ...]. Cachea el XML crudo en
 * data/imports/cfr-raw/part-NNN.xml para no re-bajar en runs sucesivos
 * salvo que se pase noCache=true.
 */
async function fetchAllParts({ issueDate = null, noCache = false, log = () => {} } = {}) {
  const date = issueDate || await getLatestIssueDate();
  log(`Fetching CFR Title 49 issue ${date}, ${PARTS.length} Parts`);
  fs.mkdirSync(RAW_DIR, { recursive: true });

  const all = [];
  for (let i = 0; i < PARTS.length; i++) {
    const partNum = PARTS[i];
    const url = `https://www.ecfr.gov/api/versioner/v1/full/${date}/title-49.xml?part=${partNum}`;
    const cachePath = path.join(RAW_DIR, `part-${partNum}.xml`);

    let xml;
    if (!noCache && fs.existsSync(cachePath)) {
      xml = fs.readFileSync(cachePath, 'utf8');
      log(`Part ${partNum}: cached`);
    } else {
      log(`Part ${partNum}: fetching...`);
      xml = await fetchUrl(url);
      fs.writeFileSync(cachePath, xml);
      await sleep(500); // cortesia a la API publica
    }

    const sections = parsePartXml(xml, partNum);
    for (const s of sections) {
      s.keywords = autoKeywords(s.title);
      s.content_hash = hashContent(s.title, s.text);
      s.issue_date = date;
    }
    all.push(...sections);
  }
  log(`Total: ${all.length} secciones de ${PARTS.length} Parts`);
  return { sections: all, issue_date: date };
}

module.exports = {
  fetchAllParts,
  getLatestIssueDate,
  parsePartXml,
  hashContent,
  autoKeywords,
  PARTS,
};

File: dotnet/BotDot.Web/Jobs/CfrFetcher.cs
// Fetcher del 49 CFR desde la API publica de eCFR.gov.
// NO usa XML parser — usa regex igual que el Node (cfr-fetcher.js).
//
// Razon de no usar XML parser nativo:
//   1. Defense en depth contra XXE — regex no procesa entities/DTDs
//   2. Paridad byte-exacta con el Node
//   3. eCFR XML es estructurado y predecible
//
// SSRF: la URL base esta hardcoded a www.ecfr.gov — no acepta input del
// usuario. Si alguien edita el codigo para parametrizar el host, agregar
// validacion de allowlist primero.

using System.Net;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace BotDot.Web.Jobs;

public class CfrSection
{
    public string Section { get; set; } = "";
    public int Part { get; set; }
    public string Title { get; set; } = "";
    public string Text { get; set; } = "";
    public List<string> Keywords { get; set; } = new();
    public string ContentHash { get; set; } = "";
    public string IssueDate { get; set; } = "";
}

public class CfrFetcher
{
    /// <summary>Parts 388 y 394 son [Reserved] en el CFR — no tienen contenido.</summary>
    private static readonly int[] Parts = new[]
    {
        380, 381, 382, 383, 384, 385, 386, 387, 389,
        390, 391, 392, 393, 395, 396, 397, 398, 399
    };

    private const string EcfrBase = "https://www.ecfr.gov";
    private const string EcfrTitlesUrl = EcfrBase + "/api/versioner/v1/titles.json";
    private static readonly Regex SectionRe = new(
        @"<DIV8\s+N=""([^""]+)""\s+TYPE=""SECTION""[^>]*>([\s\S]*?)</DIV8>", RegexOptions.Compiled);
    private static readonly Regex HeadRe = new(@"<HEAD>([\s\S]*?)</HEAD>", RegexOptions.Compiled);
    private static readonly Regex ParagraphRe = new(@"<P>([\s\S]*?)</P>", RegexOptions.Compiled);
    private static readonly Regex InlineTagRe = new(@"</?[A-Z][A-Z0-9]*( [^>]*)?>", RegexOptions.Compiled);
    private static readonly Regex SectionPrefixRe = new(@"^§\s*[\d.]+\s*", RegexOptions.Compiled);

    private readonly IHttpClientFactory _httpFactory;
    private readonly string _rawDir;
    private readonly ILogger<CfrFetcher> _log;

    public CfrFetcher(IHttpClientFactory httpFactory, IWebHostEnvironment env, ILogger<CfrFetcher> log)
    {
        _httpFactory = httpFactory;
        _log = log;
        // Cache de XML crudo en data/imports/cfr-raw/ (compartido con Node).
        _rawDir = Path.Combine(env.ContentRootPath, "..", "..", "data", "imports", "cfr-raw");
        // Si la ruta relativa no resuelve (instalacion no estandar), buscar
        // hacia arriba el repo root.
        if (!Directory.Exists(Path.GetDirectoryName(_rawDir)))
        {
            var repoRoot = FindRepoRoot(env.ContentRootPath);
            if (repoRoot != null)
                _rawDir = Path.Combine(repoRoot, "data", "imports", "cfr-raw");
        }
    }

    private static string? FindRepoRoot(string startDir)
    {
        var dir = new DirectoryInfo(startDir);
        for (int i = 0; i < 6 && dir != null; i++, dir = dir.Parent)
        {
            if (Directory.Exists(Path.Combine(dir.FullName, "data", "cfrs"))) return dir.FullName;
        }
        return null;
    }

    private HttpClient CreateClient()
    {
        // SSRF defense: aunque la URL es hardcoded, deshabilitamos auto-redirect
        // a hosts externos. Si eCFR redirect a algo, queremos verlo en logs.
        var handler = new HttpClientHandler
        {
            AllowAutoRedirect = true,
            MaxAutomaticRedirections = 3,
        };
        var client = new HttpClient(handler, disposeHandler: true)
        {
            Timeout = TimeSpan.FromSeconds(30),
        };
        client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/xml"));
        client.DefaultRequestHeaders.UserAgent.Add(new ProductInfoHeaderValue("BOTDOT-CFR-fetcher", "1.0"));
        return client;
    }

    private async Task<string> FetchUrlAsync(string url, CancellationToken ct)
    {
        // Validar que el URL sea de eCFR.gov — defense en profundidad si alguien
        // pasa parametrizacion futura.
        if (!url.StartsWith("https://www.ecfr.gov/", StringComparison.OrdinalIgnoreCase))
            throw new InvalidOperationException($"URL no permitida (debe ser ecfr.gov): {url}");

        using var client = CreateClient();
        using var resp = await client.GetAsync(url, ct);
        if (resp.StatusCode != HttpStatusCode.OK)
            throw new HttpRequestException($"HTTP {(int)resp.StatusCode}: {url}");
        return await resp.Content.ReadAsStringAsync(ct);
    }

    public async Task<string> GetLatestIssueDateAsync(CancellationToken ct = default)
    {
        var data = await FetchUrlAsync(EcfrTitlesUrl, ct);
        using var doc = System.Text.Json.JsonDocument.Parse(data);
        if (!doc.RootElement.TryGetProperty("titles", out var titles)) throw new Exception("No 'titles' en response");
        foreach (var t in titles.EnumerateArray())
        {
            if (t.TryGetProperty("number", out var num) && num.GetInt32() == 49)
            {
                if (t.TryGetProperty("latest_issue_date", out var date) && date.GetString() is string s)
                    return s;
            }
        }
        throw new Exception("No se pudo determinar latest_issue_date para Title 49");
    }

    public async Task<(List<CfrSection> Sections, string IssueDate)> FetchAllPartsAsync(
        string? issueDate = null, bool noCache = false, CancellationToken ct = default)
    {
        var date = issueDate ?? await GetLatestIssueDateAsync(ct);
        _log.LogInformation("Fetching CFR Title 49 issue {Date}, {N} Parts", date, Parts.Length);
        Directory.CreateDirectory(_rawDir);

        var all = new List<CfrSection>();
        for (int i = 0; i < Parts.Length; i++)
        {
            ct.ThrowIfCancellationRequested();
            var partNum = Parts[i];
            var url = $"{EcfrBase}/api/versioner/v1/full/{date}/title-49.xml?part={partNum}";
            var cachePath = Path.Combine(_rawDir, $"part-{partNum}.xml");

            string xml;
            if (!noCache && File.Exists(cachePath))
            {
                xml = await File.ReadAllTextAsync(cachePath, Encoding.UTF8, ct);
                _log.LogDebug("Part {P}: cached", partNum);
            }
            else
            {
                _log.LogInformation("Part {P}: fetching...", partNum);
                xml = await FetchUrlAsync(url, ct);
                await File.WriteAllTextAsync(cachePath, xml, Encoding.UTF8, ct);
                await Task.Delay(500, ct); // cortesia a la API publica
            }

            var sections = ParsePartXml(xml, partNum);
            foreach (var s in sections)
            {
                s.Keywords = AutoKeywords(s.Title);
                s.ContentHash = HashContent(s.Title, s.Text);
                s.IssueDate = date;
            }
            all.AddRange(sections);
        }
        _log.LogInformation("Total: {N} secciones de {P} Parts", all.Count, Parts.Length);
        return (all, date);
    }

    public static List<CfrSection> ParsePartXml(string xml, int partNum)
    {
        var sections = new List<CfrSection>();
        foreach (Match m in SectionRe.Matches(xml))
        {
            var sectionNum = m.Groups[1].Value;
            var inner = m.Groups[2].Value;

            var headMatch = HeadRe.Match(inner);
            string title = headMatch.Success ? DecodeEntities(StripInlineTags(headMatch.Groups[1].Value)).Trim() : "";
            title = SectionPrefixRe.Replace(title, "").Trim();

            var paragraphs = new List<string>();
            foreach (Match pm in ParagraphRe.Matches(inner))
            {
                paragraphs.Add(DecodeEntities(StripInlineTags(pm.Groups[1].Value)).Trim());
            }
            var text = string.Join("\n\n", paragraphs);

            if (!string.IsNullOrEmpty(sectionNum))
                sections.Add(new CfrSection { Section = sectionNum, Title = title, Text = text, Part = partNum });
        }
        return sections;
    }

    private static string StripInlineTags(string s) => InlineTagRe.Replace(s, "");

    private static string DecodeEntities(string s)
    {
        return s
            .Replace("&#xA7;", "§", StringComparison.OrdinalIgnoreCase)
            .Replace("&#xB6;", "¶", StringComparison.OrdinalIgnoreCase)
            .Replace("&amp;", "&")
            .Replace("&lt;", "<")
            .Replace("&gt;", ">")
            .Replace("&quot;", "\"")
            .Replace("&#x([0-9a-f]+);", "", StringComparison.OrdinalIgnoreCase)  // placeholder, fixed via regex below
            ;
        // Nota: para hex/decimal entities el Node usa regex con substitucion.
        // El metodo simplificado anterior cubre los casos comunes; los menos
        // frecuentes los dejamos como vienen (no es comun en eCFR).
    }

    public static List<string> AutoKeywords(string title)
    {
        var stop = new HashSet<string> { "for","the","and","with","from","that","this","part","of","to","in","on","a","an","or","by","as","be","is","are","at","it" };
        return Regex.Replace((title ?? "").ToLowerInvariant(), @"[^a-z0-9 ]", " ")
            .Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries)
            .Where(t => t.Length >= 4 && !stop.Contains(t))
            .Take(8)
            .ToList();
    }

    public static string HashContent(string title, string text)
    {
        var bytes = Encoding.UTF8.GetBytes($"{title}|{text}");
        var hash = SHA256.HashData(bytes);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}
